File: lib/src/main/java/io/xj/lib/meme/MemeTaxonomy.java
package io.xj.lib.meme;

import java.util.Map;
import io.xj.lib.util.StringUtils;

import org.jetbrains.annotations.Nullable;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Locale;
import java.util.Objects;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * TemplateConfig has Meme categories
 * https://www.pivotaltracker.com/story/show/181801646
 * <p>
 * <p>
 * A template configuration has a field called `memeTaxonomy` which defines the taxonomy of memes.
 * <p>
 * For example, this might look like
 * <p>
 * ```
 * memeTaxonomy=CITY[CHICAGO,DENVER,PHILADELPHIA]
 * ```
 * <p>
 * That would tell XJ about the existence of a meme category called City with values `CHICAGO`, `DENVER`, and `PHILADELPHIA`. And these would function as exclusion like numeric memes, e.g. after content having `CHICAGO` is chosen, we can choose nothing with `DENVER` or `PHILADELPHIA`.
 */
public class MemeTaxonomy {
  static final String CATEGORY_SEPARATOR = ";";
  final List<Category> categories;

  MemeTaxonomy() {
    categories = List.of();
  }

  MemeTaxonomy(@Nullable String raw) {
    if (StringUtils.isNullOrEmpty(raw))
      categories = List.of();
    else
      categories = Arrays.stream(Objects.requireNonNull(raw).split(CATEGORY_SEPARATOR))
        .map(Category::new)
        .filter(Category::hasMemes)
        .toList();
  }

  public MemeTaxonomy(List<Object> data) {
    categories = data.stream()
      .flatMap(d -> {
        if (d instanceof Map<?, ?>) try {
          //noinspection unchecked
          return Stream.of(Category.fromMap((Map<String, Object>) d));
        } catch (Exception ignored) {
          //noop
        }
        return Stream.empty();
      })
      .collect(Collectors.toList());
  }

  public static MemeTaxonomy fromString(@Nullable String raw) {
    return new MemeTaxonomy(raw);
  }

  public static MemeTaxonomy fromList(List<Object> data) {
    return new MemeTaxonomy(data);
  }

  public static MemeTaxonomy empty() {
    return new MemeTaxonomy();
  }

  public String toString() {
    return categories.stream()
      .map(Category::toString)
      .collect(Collectors.joining(CATEGORY_SEPARATOR));
  }

  public List<Map<String, Object>> toList() {
    return categories.stream()
      .map(Category::toMap)
      .toList();
  }

  public List<Category> getCategories() {
    return categories;
  }

  public boolean isAllowed(List<String> memes) {
    return categories.stream().allMatch(category -> category.isAllowed(memes));
  }

  public static class Category {
    static final Pattern rgx = Pattern.compile("^([a-zA-Z ]+)\\[([a-zA-Z, ]+)]$");
    static final String MEME_SEPARATOR = ",";
    static final String KEY_NAME = "name";
    static final String KEY_MEMES = "memes";
    static final String DEFAULT_CATEGORY_NAME = "CATEGORY";
    final String name;
    final List<String> memes;

    Category(@Nullable String raw) {
      if (StringUtils.isNullOrEmpty(raw)) {
        name = DEFAULT_CATEGORY_NAME;
        memes = List.of();
        return;
      }

      Matcher matcher = rgx.matcher(Objects.requireNonNull(raw).trim());

      if (!matcher.find()) {
        name = DEFAULT_CATEGORY_NAME;
        memes = List.of();
        return;
      }

      String pfx = matcher.group(1);
      if (java.util.Objects.isNull(pfx) || pfx.length() == 0) {
        name = DEFAULT_CATEGORY_NAME;
        memes = List.of();
        return;
      }
      name = sanitize(StringUtils.toAlphabetical(pfx));

      String body = matcher.group(2);
      if (java.util.Objects.isNull(body) || body.length() == 0)
        memes = List.of();
      else
        memes = Arrays.stream(body.split(MEME_SEPARATOR))
          .map(this::sanitize)
          .toList();
    }

    public Category(Map<String, Object> data) {
      name = data.containsKey(KEY_NAME) ? sanitize(String.valueOf(data.get(KEY_NAME))) : DEFAULT_CATEGORY_NAME;
      memes = parseMemeList(data);
    }

    List<String> parseMemeList(Map<String, Object> data) {
      if (data.containsKey(KEY_MEMES) && data.get(KEY_MEMES) instanceof List<?>) try {
        //noinspection unchecked
        return ((List<String>) data.get(KEY_MEMES)).stream().map(this::sanitize).toList();
      } catch (Exception ignored) {
        //noop
      }
      return List.of();
    }

    public static Category fromString(@Nullable String raw) {
      return new Category(raw);
    }

    public static Category fromMap(Map<String, Object> data) {
      return new Category(data);
    }

    public String toString() {
      return String.format("%s[%s]", name, String.join(MEME_SEPARATOR, memes));
    }

    public String getName() {
      return name;
    }

    public List<String> getMemes() {
      return memes;
    }

    public boolean isAllowed(List<String> targets) {
      return 1 >= new HashSet<>(targets)
        .stream()
        .filter(memes::contains)
        .count();
    }

    public boolean hasMemes() {
      return !memes.isEmpty();
    }

    public Map<String, Object> toMap() {
      return Map.of(
        KEY_NAME, name,
        KEY_MEMES, memes
      );
    }

    String sanitize(String raw) {
      return StringUtils.toAlphabetical(raw.trim()).toUpperCase(Locale.ROOT);
    }
  }
}

File: engine/src/meme/MemeTaxonomy.cpp
// Copyright (c) XJ Music Inc. (https://xjmusic.com) All Rights Reserved.

#include <sstream>
#include <regex>
#include <set>

#include <spdlog/spdlog.h>

#include "xjmusic/meme/MemeTaxonomy.h"

using namespace XJ;


std::string MemeCategory::MEME_SEPARATOR = ",";
std::string MemeCategory::KEY_NAME = "name";
std::string MemeCategory::KEY_MEMES = "memes";
std::string MemeCategory::DEFAULT_CATEGORY_NAME = "CATEGORY";
std::regex MemeCategory::rgx("^([a-zA-Z ]+)\\[([a-zA-Z, ]+)\\]$");


static std::string sanitize(const std::string &raw) {
  return StringUtils::toUpperCase(StringUtils::toAlphabetical(StringUtils::trim(raw)));
}


std::set<std::string>
MemeCategory::parseMemeList(const MapStringToOneOrManyString &data) {
  const auto it = data.find(KEY_MEMES);
  if (it != data.end() && std::holds_alternative<std::set<std::string>>(it->second)) {
    const std::set<std::string> rawList = std::get<std::set<std::string>>(it->second);
    std::set<std::string> sanitizedList;
    for (const auto &raw: rawList) {
      sanitizedList.insert(sanitize(raw));
    }
    return sanitizedList;
  }
  return {};
}


MemeCategory::MemeCategory(const std::string *raw) {
  if (StringUtils::isNullOrEmpty(raw)) {
    name = DEFAULT_CATEGORY_NAME;
    memes = std::set<std::string>();
    return;
  }

  std::smatch matcher;
  std::regex_search(*raw, matcher, rgx);

  if (matcher.empty()) {
    name = DEFAULT_CATEGORY_NAME;
    memes = std::set<std::string>();
    return;
  }

  std::string pfx = matcher[1].str();
  if (pfx.empty()) {
    name = DEFAULT_CATEGORY_NAME;
    memes = std::set<std::string>();
    return;
  }
  std::string alphabetical = StringUtils::toAlphabetical(pfx);
  name = sanitize(alphabetical);

  std::string body = matcher[2].str();
  if (body.empty())
    memes = std::set<std::string>();
  else {
    memes.clear();
    std::istringstream iss(body);
    for (std::string s; std::getline(iss, s, ',');)
      memes.insert(sanitize(s));
  }
}


MemeCategory::MemeCategory(const MapStringToOneOrManyString &data) {
  const auto it = data.find(KEY_NAME);
  name = (it != data.end() && std::holds_alternative<std::string>(it->second))
         ? sanitize(std::get<std::string>(it->second))
         : DEFAULT_CATEGORY_NAME;

  memes = parseMemeList(data);
}


std::string MemeCategory::getName() const {
  return name;
}


std::set<std::string> MemeCategory::getMemes() {
  return memes;
}


bool MemeCategory::isAllowed(std::set<std::string> &targets) const {
  int count = 0;
  for (const auto &meme: memes) {
    if (targets.find(meme) != targets.end()) {
      count++;
    }
  }
  return count <= 1;
}


bool MemeCategory::hasMemes() const {
  return !memes.empty();
}


std::string MemeCategory::toString() const {
  std::vector<std::string> sortedMemes(memes.begin(), memes.end());
  std::sort(sortedMemes.begin(), sortedMemes.end());
  return name + "[" + StringUtils::join(sortedMemes, MEME_SEPARATOR) + "]";
}


MapStringToOneOrManyString MemeCategory::toMap() const {
  return MapStringToOneOrManyString{
      {KEY_NAME,  name},
      {KEY_MEMES, memes}};
}


char MemeTaxonomy::CATEGORY_SEPARATOR = ';';


MemeTaxonomy::MemeTaxonomy(const std::string &raw) : MemeTaxonomy() {
  if (StringUtils::isNullOrEmpty(&raw)) {
    return;
  }
  std::istringstream iss(raw);
  for (std::string s; std::getline(iss, s, CATEGORY_SEPARATOR);) {
    categories.insert(MemeCategory(&s));
  }
}


MemeTaxonomy::MemeTaxonomy(const std::set<MapStringToOneOrManyString> &data) {
  categories.clear();
  for (auto &d: data) {
    try {
      categories.insert(MemeCategory(d));
    } catch (...) {
      spdlog::error("Failed to add map data!");
    }
  }
}


std::string MemeTaxonomy::toString() const {
  std::vector<std::string> strings;
  strings.reserve(categories.size());
  for (const auto &category: categories) {
    strings.push_back(category.toString());
  }
  std::sort(strings.begin(), strings.end());
  return StringUtils::join(strings, std::string(1, CATEGORY_SEPARATOR));
}


std::set<MapStringToOneOrManyString> MemeTaxonomy::toList() const {
  std::set<MapStringToOneOrManyString> data;
  for (const auto &category: categories) {
    auto map = category.toMap();
    data.insert(map);
  }
  return data;
}


std::set<MemeCategory> MemeTaxonomy::getCategories() {
  return categories;
}


bool MemeTaxonomy::isAllowed(std::set<std::string> memes) const {
  for (const auto &memeCategory: categories) {
    if (!memeCategory.isAllowed(memes)) {
      return false;
    }
  }
  return true;
}


MemeTaxonomy MemeTaxonomy::empty() {
  return {};
}


MemeTaxonomy MemeTaxonomy::fromSet(std::set<MapStringToOneOrManyString> &data) {
  return MemeTaxonomy(data);
}


MemeTaxonomy MemeTaxonomy::fromString(const std::string &raw) {
  return MemeTaxonomy(raw);
}


MemeTaxonomy MemeTaxonomy::fromList(
    const std::vector<std::map<std::string, std::variant<std::string, std::vector<std::string>>>> &list) {
  std::set<MapStringToOneOrManyString> set;
  for (const auto &map: list) {
    MapStringToOneOrManyString m;
    for (const auto &[key, value]: map) {
      if (std::holds_alternative<std::vector<std::string>>(value)) {
        auto v = std::get<std::vector<std::string>>(value);
        m[key] = std::set<std::string>(v.begin(), v.end());
      } else {
        m[key] = std::get<std::string>(value);
      }
    }
    set.insert(m);
  }
  return MemeTaxonomy(set);
}

